go/src/koding/klient/machine/mount/sync.go

```diff
@@ -0,0 +1,301 @@
+// NewSync creates a new Sync instance from the given options.
+func NewSync(opts SyncOpts) (*Sync, error) {
+	if err := opts.Valid(); err != nil {
+		return nil, err
+	}
+
+	s := &Sync{
+		opts:  opts,
+		syncs: make(map[ID]*synced),
+	}
+
+	if opts.Log != nil {
+		s.log = opts.Log.New("sync")
+	} else {
+		s.log = machine.DefaultLogger.New("sync")
+	}
+
+	return s, nil
+}
+
+// Add starts synchronization between remote and local directories. It creates
+// all necessary cache files if they are not present.
+func (s *Sync) Add(id ID, m Mount) error {
+	s.mu.RLock()
+	sd, ok := s.syncs[id]
+	s.mu.RUnlock()
+
+	if ok {
+		return fmt.Errorf("mount with ID %s already exists (%s)", id, sd.m)
+	}
+
+	sd, err := newSynced(id, m, &s.opts)
+	if err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	if _, ok := s.syncs[id]; ok {
+		s.mu.Unlock()
+		return fmt.Errorf("mount with ID %s added twice", id)
+	}
+	s.syncs[id] = sd
+	s.mu.Unlock()
+
+	return nil
+}
+
+// Info returns the current state of mount with provided ID.
+func (s *Sync) Info(id ID) (*Info, error) {
+	s.mu.RLock()
+	sd, ok := s.syncs[id]
+	s.mu.RUnlock()
+
+	if !ok {
+		return nil, ErrMountNotFound
+	}
+
+	return sd.info(), nil
+}
+
+// Drop removes the mount and clean ups the resources it uses.
+func (s *Sync) Drop(id ID) (err error) {
+	s.mu.Lock()
+	sd, ok := s.syncs[id]
+	delete(s.syncs, id)
+	s.mu.Unlock()
+
+	if !ok {
+		return nil
+	}
+
+	if err = sd.drop(); err != nil {
+		// Drop failed - put synced back to the map.
+		s.mu.Lock()
+		s.syncs[id] = sd
+		s.mu.Unlock()
+	}
+
+	return err
+}
+
+// synced stores and synchronizes single mount. The main goal of synced logic
+// is to make remote and local indexes similar.
+type synced struct {
+	opts *SyncOpts
+	id   ID     // identifier of synced mount.
+	m    Mount  // single mount with absolute paths.
+	wd   string // working directory of the mount.
+
+	ridx *index.Index // known state of remote index.
+	lidx *index.Index // known state of local index.
+}
+
+// newSynced creates a new sync instance. It ensures basic mount directory
+// structure. This function is blocking.
+func newSynced(id ID, m Mount, opts *SyncOpts) (*synced, error) {
+	s := &synced{
+		opts: opts,
+		id:   id,
+		m:    m,
+		wd:   filepath.Join(opts.WorkDir, "mount-"+string(id)),
+	}
+
+	// Create directory structure if it doesn't exist.
+	if err := s.mktree(); err != nil {
+		return nil, err
+	}
+
+	// Fetch or load remote index.
+	var err error
+	if s.ridx, err = s.loadIdx(RemoteIndexName, s.fetchRemoteIdx); err != nil {
+		return nil, err
+	}
+
+	// Create or load local index.
+	if s.lidx, err = s.loadIdx(LocalIndexName, s.fetchLocalIdx); err != nil {
+		return nil, err
+	}
+
+	// Update local index if needed.
+	if err := s.updateLocal(); err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// mktree ensures that synced working directory is created.
+func (s *synced) mktree() error {
+	dataPath := filepath.Join(s.wd, "data")
+	info, err := os.Stat(dataPath)
```

Comment:
prob MkdirAll is enough, it already handles Stat as fast path https://golang.org/src/os/path.go?s=537:766